gateway/ln-gateway/src/state_machine/pay.rs

```diff
@@ -183,20 +183,104 @@ impl GatewayPayInvoice {
     ) -> Vec<StateTransition<GatewayPayStateMachine>> {
         let payload = self.pay_invoice_payload.clone();
         vec![StateTransition::new(
-            Self::await_get_payment_parameters(
+            Self::fetch_parameters_and_pay(
                 global_context,
-                self.pay_invoice_payload.contract_id,
+                payload.clone(),
                 context.clone(),
+                common.clone(),
             ),
-            move |_dbtx, result, _old_state| {
-                Box::pin(Self::transition_buy_preimage(
+            move |_dbtx, result, _old_state| Box::pin(futures::future::ready(result)),
```

Comment:
The more canonical way of implementing trigger future and state transition would be to only give the result to the state transition function and then constructing the new state there. Was there any particular reason not to do it that way?

Reply:
Yeah, it's a bit unorthodox, but this was the natural way of doing it after the refactoring. The problem is that the output of this function is very complex due to the different outcomes and the best representation are the state structures  themselves.

But are there any issues doing this way? I couldn't find.

Reply:
There aren't any direct problems with it, just a bit confusing.